design_documents/cairo_deployment_scripts.md

```diff
@@ -0,0 +1,336 @@
+# Cairo deployment scripts
+
+## Table of contents
+
+* [Context](#context)
+* [Goal](#goal)
+* [Considered Solutions](#considered-solution)
+  * [sncast commands](#sncast-commands)
+  * [sncast config](#sncast-config)
+  * [interacting with contract](#interacting-with-contract)
+  * [running the script](#running-the-script)
+  * [error handling](#error-handling)
+  * [idempotency](#idempotency)
+  * [example script](#example-script)
+  * [example state file](#example-state-file)
+  * [miscellaneous](#miscellaneous)
+* [Proposed steps and tasks](#proposed-stepstasks)
+
+## Context
+
+There should be a possibility to write a script in cairo, that would enable users to make transactions and send them 
+to starknet chain. It should allow to declare and deploy contracts as well as apply state transitions on already deployed
+contracts.
+
+## Goal
+
+Propose a solution with an example syntax, that would allow to write deployment scripts in cairo.
+
+## Considered solution
+
+This section is split into smaller subsections describing things we will need to tackle while implementing the solution.
+
+### sncast commands
+Specific sncast commands (declare, deploy, account) could be imported as regular functions to the scipts, and called as such.
+Our functions return specific types (structs defined in `cast/src/helpers/response_structs.rs`), that make retrieving
+essential information easier, so we should be good on that front.
+
+The commands would have to be implemented in the same manner as forge's cheatcodes.
+
+### sncast config
+We should allow for all flags passed to sncast to be propagated to the script. CastConfig should also be importable
+and usable from script if someone wishes so, but we should not require it. That means, we might have to review our subcommands
+and if some of them requires it directly (`account create` is one), we should try to get rid of this dependency. 
+
+### interacting with contract
+We will use contracts dispatchers to be able to call/invoke its functions directly 
+(e.g. if contract named `mycontract` has a function `myfunction`, we should be able to just do `mycontract.myfunction();`).
+Later on we could also support our subcommands (`invoke`, `call`) to call/invoke contracts without dispatchers.
+
+### running the script
+The script would essentially run an entrypoint function - `main`. Inside our script subcommand, we will have to compile 
+the cairo script file using `scarb build`, and then run it using `scarb cairo-run`. 
```

Comment:
Is this going to be doable? Custom hint logic will require us to probably handle both compilation and running

Reply:
Yes, I had a meeting with Artur and he outlined to me how it works now in forge (which is a bit different than I imagined). We'll be doing custom build/extension and a runner similarly to forge. Updated this section in newest commit. Sorry I didn't do it sooner.